design/backup-performance-improvements.md

```diff
@@ -0,0 +1,357 @@
+- Phase 2 will add multiple worker threads for backing up item blocks, so instead of backing up each block as it identified, the velero backup workflow will instead add the block to a channel and one of the workers will pick it up.
+- Actual support for VolumeGroupSnapshots is out-of-scope here and will be handled in a future design proposal, but the item block refactor introduced in Phase 1 is a primary building block for this future proposal.
+
+## Background
+Currently, during backup processing, the main Velero backup controller runs in a single thread, completely finishing the primary backup processing for one resource before moving on to the next one.
+We can improve the overall backup performance by backing up multiple items for a backup at the same time, but before we can do this we must first identify resources that need to be backed up together.
+Generally speaking, resources that need to be backed up together are resources with interdependencies -- pods with their PVCs, PVCs with their PVs, groups of pods that form a single application, CRs, pods, and other resources that belong to the same operator, etc.
+As part of this initial refactoring, once these "Item Blocks" are identified, an additional change will be to move pod hook processing up to the ItemBlock level.
+If there are multiple pods in the ItemBlock, pre-hooks for all pods will be run before backing up the items, followed by post-hooks for all pods.
+This change to hook processing is another prerequisite for future VolumeGroupSnapshot support, since supporting this will require backing up the pods and volumes together for any volumes which belong to the same group.
+Once we are backing up items by block, the next step will be to create multiple worker threads to process and back up ItemBlocks, so that we can back up multiple ItemBlocks at the same time.
+
+In looking at the different kinds of large backups that Velero must deal with, two obvious scenarios come to mind:
+1. Backups with a relatively small number of large volumes
+2. Backups with a large number of relatively small volumes.
+
+In case 1, the majority of the time spent on the backup is in the asynchronous phases -- CSI snapshot creation actions after the snaphandle exists, and DataUpload processing. In that case, parallel item processing will likely have a minimal impact on overall backup completion time.
+
+In case 2, the majority of time spent on the backup will likely be during the synchronous actions. Especially as regards CSI snapshot creation, the waiting for the VSC snaphandle to exist will result in significant passage of time with thousands of volumes. This is the sort of use case which will benefit the most from parallel item processing.
+
+## Goals
+- Identify groups of related items to back up together (ItemBlocks).
+- Manage backup hooks at the ItemBlock level rather than per-item.
+- Using worker threads, back up ItemBlocks at the same time.
+
+## Non Goals
```

Comment:
Do you think restore is also something for which we can improve performance with the approach in this doc in the future.
Or is it strictly non-goal for now.

Reply:
Restore is definitely out-of-scope here. While we may be able to adopt a similar approach to restore in the future, it would be a completely different set of work -- some of this code may end up being refactored to help support restore as well, but by and large, the restore work is in addition to this work, so it makes sense to do that in the future as a separate task. We won't be saving ourselves much work by trying to do it all at once.

Also, it's a lower priority for a couple reasons:
1) Backups happen far more often than restores, so reducing backup time will have more of an impact on overall velero load and daily tasks.
2) Outside of PV content restore (which is already done in parallel) the biggest bottleneck for restores right now is the per-resource APIServer calls to Create (and then Update) each item. I don't anticipate a big reduction in restore time by running parallel threads within a restore. 

I'll add Restore to non-goals.

Reply:
If we make per resource API calls in parallel. Won't that reduce the restore time?

Reply:
@anshulahuja98 it depends on where the bottleneck is. In any case I'd expect less benefit here than on backup. I'm not saying we never want to do restore -- but this is already a big task split into two phases with a lot of changes packed into this design. Restore would be another big chunk of work with its own design. We can consider that in a future proposal.

Reply:
Alright
I understand that backup is the priority here. And that makes sense. 
Just wanted to understand the overall thought process.


One more closing thought, as part of backup we'll make some changes in the hooks execution as part of item block.
Do you think this will have touchpoints with the restore hooks implementation.

Reply:
@anshulahuja98 Even if we don't end up making restores work with parallel threads, we may end up needing to make similar changes to hooks (and possibly using the ItemBlock concept) on the restore side when we implement VolumeGroupSnapshot support. On the backup side, VolumeGroupSnapshot support will involve grouping all of the pods mounting volumes in the same VolumeGroup into the same ItemBlock, so hooks will run before and after the VolumeGroupSnapshot is created. I imagine we'll want similar changes on the restore side. Whether this means the same ItemBlock struct will be enhanced to work for both backup and restore or we need a similar-but-not-identical struct on the restore side isn't totally clear yet -- that will need to be worked out when we dig into the VolumeGroupSnapshot design work.

Reply:
@anshulahuja98 I think you're right regarding the parallel calls to the API server,up hitting a new bottleneck around priority and fairness. Dealing with that alone should require an entirely sepa but you do end rate design for that reason.

Reply:
We recently hit the following issue - https://github.com/vmware-tanzu/velero/issues/7991
So throttling a top of mind for us, anything we can do even for backup side to have sake of mind might be useful.

Reply:
@anshulahuja98 I think on the backup side, this concern is limited to plugin actions. Backup process itself has already made all of the List calls and built up the backup list. Plugins might make additional APIServer calls, and any additional items returned from them might need to be retrieved, but those are really the only ones we expect here -- so in almost all cases, I imagine this will be many fewer APIServer calls than a regular restore.

Reply:
@anshulahuja98 in any case, the above throttling/error situation doesn't seem to be exactly the situation of "too many APIServer calls "Error: client rate limiter Wait returned an error: rate: Wait(n=1) would exceed context deadline" -- this could be timeout-related, or it could be a result of throttling outside of Velero completely. In my experience with Velero and making many APIServer calls, the throttling on the client side tends to  be silent. QPS is hit, and the client simply limits how many calls it makes. If we're overloading the APIServer without hitting the QPS limit, then the answer may be to lower the QPS setting. @shawn-hurley do you have any input here -- do we need to do anything different in our backup design to handle throttling? I'm not currently seeing anything obvious we need to add here, but I may be missing something.

Reply:
Wil resolve this one for now, would request @shawn-hurley  to also pitch in if he has any thoughts.